src/averaging.rs

```diff
@@ -0,0 +1,303 @@
+    );
+
+    // iterate through the time dimension of the arrays in chunks of size `time_factor`.
+    for (
+        jones_timestep_chunk,
+        weight_timestep_chunk,
+        flag_timestep_chunk,
+        mut averaged_jones_timestep_view,
+        mut averaged_weight_timestep_view,
+        mut averaged_flag_timestep_view,
+    ) in izip!(
+        jones_array.axis_chunks_iter(Axis(0), time_factor),
+        weight_array.axis_chunks_iter(Axis(0), time_factor),
+        flag_array.axis_chunks_iter(Axis(0), time_factor),
+        averaged_jones_array.outer_iter_mut(),
+        averaged_weight_array.outer_iter_mut(),
+        averaged_flag_array.outer_iter_mut(),
+    ) {
+        // iterate through the channel dimension of the arrays in chunks of size `frequency_factor`.
+        for (
+            jones_channel_chunk,
+            weight_channel_chunk,
+            flag_channel_chunk,
+            mut averaged_jones_channel_view,
+            mut averaged_weight_channel_view,
+            mut averaged_flag_channel_view,
+        ) in izip!(
+            jones_timestep_chunk.axis_chunks_iter(Axis(1), frequency_factor),
+            weight_timestep_chunk.axis_chunks_iter(Axis(1), frequency_factor),
+            flag_timestep_chunk.axis_chunks_iter(Axis(1), frequency_factor),
+            averaged_jones_timestep_view.outer_iter_mut(),
+            averaged_weight_timestep_view.outer_iter_mut(),
+            averaged_flag_timestep_view.outer_iter_mut(),
+        ) {
+            // iterate through the baseline dimension of the arrays.
+            for (
+                jones_chunk,
+                weight_chunk,
+                flag_chunk,
+                mut averaged_jones_view,
+                mut averaged_weight_view,
+                mut averaged_flag_view,
+            ) in izip!(
+                jones_channel_chunk.axis_iter(Axis(2)),
+                weight_channel_chunk.axis_iter(Axis(2)),
+                flag_channel_chunk.axis_iter(Axis(2)),
+                averaged_jones_channel_view.outer_iter_mut(),
+                averaged_weight_channel_view.outer_iter_mut(),
+                averaged_flag_channel_view.outer_iter_mut(),
+            ) {
+                // dimensions:
+                // - jones_chunk -> [tf, ff] (each element is a complex vis array of length pol)
+                // - weight_chunk -> [tf, ff, pol]
+                // - flag_chunk shape -> [tf, ff, pol]
+                // - averaged_jones_view -> []
+                // - averaged_weight_view -> [pol]
+                // - averaged_flag_view -> [pol]
+
+                // for each time, frequency and pol in the chunks, compute the
+                // weighted geometric average of unflagged visibilities.
+
+                let chunk_size = jones_chunk.dim().0 * jones_chunk.dim().1;
+
+                for pol_idx in 0..4 {
+                    let mut all_flagged = true;
+                    let mut pol_sum: Complex<f32> = Complex::new(0., 0.);
```

Comment:
Can we accumulate into a c64 instead? It seems like promoting and demoting the data has very little cost and we'll save a bit of precision (paramount for EoR)